Implement basic equivocations detection loop

serban300 opened the pull request:
Related to #2279

First version of the equivocation detection loop. 

Known limitations that will be addressed in future PRs:
- the loop is just implemented, but it's not used anywhere. We need to start it together with the finality loop probably
- no unit tests
- limited error management: in case of an error, we log it, try to reconnect and skip the item. Maybe we should retry
- no integration test

acatangiu commented on relays/equivocation/src/reporter.rs:
```diff
@@ -0,0 +1,78 @@
+		Ok(())
+	}
+
+	fn do_process_pending_reports(&mut self, cx: &mut Context<'_>) -> Poll<()> {
```
nit: this looks like a `impl Poll for EquivocationsReporter` model (instead of `pub async fn process_pending_reports()`)

serban300 replied:
I think using `impl Future for EquivocationsReporter` could be confusing. Because if we consider `EquivocationsReporter` as a `Future`, probably normally `poll` should return `Poll::Pending` if there are pending equivocation reports. But we don't want this, because we want to be able to perform other checks while waiting for the pending reports.

That's why I just defined a `do_process_pending_reports()` method that terates through the reports, and then returns Poll::Ready().

acatangiu commented on relays/equivocation/src/equivocation_loop.rs:
```diff
@@ -0,0 +1,327 @@
+};
+use std::{future::Future, time::Duration};
+
+struct EquivocationReportingContext<P: EquivocationDetectionPipeline> {
```
Could do with some descriptive docs for all structs and methods throughout this file

acatangiu commented on relays/equivocation/src/equivocation_loop.rs:
```diff
@@ -0,0 +1,327 @@
+
+			// Check the next block
+			self.update_until_block_num().await;
+			if self.from_block_num <= self.until_block_num {
+				let mut context = match self.build_context(self.from_block_num).await {
+					Some(context) => context,
+					None => return,
+				};
+
+				self.check_block(self.from_block_num, &mut context).await;
+
+				self.from_block_num = self.from_block_num.saturating_add(1.into());
+			}
```
This could be a loop `while self.from_block_num <= self.until_block_num {}` and do full catch-up checks in a single tick.

Doesn't look like that much work to have latency concerns..

Also I would actually suggest to skip the checking full (sub)chain and just check latest (source highest/best finalized as seen by target) - only if that has different hash than block on source at same height, then you know it forked somewhere since last tick and you can iterate to find the fork. WDYT?

acatangiu replied:
> I think an attacker(s) may be able to submit `{ submit_finality_proof(forked_header#100), submit_message(malicious_message), submit_finality_proof(good_header#101) }` in a single block. So if you'll just look at the latest, you may skip this forked_header submission. So imo we should look at all headers here

but how would `good_header#101` have the correct hash if it's been built on top of `bad_header#100`? Once relayer introduces any forked header, all subsequent child chain will be a fork, no?

svyatonik replied:
We don't check the connection between headers anywhere. So it could be built on top of other "normal" header. E.g. in following schema:
```
99 ---> 100 ---> 101
     \--> 100' 
```
the relayer1 may track the bad fork and submit the `100'` first and relayer2 may then submit the `101`

acatangiu replied:
> We don't check the connection between headers anywhere.

oh, ok, I thought (without checking the code) we did.

valid scenario then, let's keep the equivocation check for every submitted header 👍

acatangiu reviewed (commented):
Looks good so far, left some comments

svyatonik commented on relays/lib-substrate-relay/src/equivocation/target.rs:
```diff
@@ -68,6 +68,12 @@ impl<P: SubstrateEquivocationDetectionPipeline, TargetClnt: Client<P::TargetChai
 	TargetClient<EquivocationDetectionPipelineAdapter<P>>
 	for SubstrateEquivocationTarget<P, TargetClnt>
 {
+	async fn best_finalized_header_number(
+		&self,
+	) -> Result<BlockNumberOf<P::TargetChain>, Self::Error> {
```
It returns best finalized target chain block, which is the target chain property - it doesn't access the runtime code or storage at all. Other methods are working with the target chain runtime code/storage/state, which was active `at` some target block. So this method returns an `anchor` that is later used as `at` in other method calls

serban300 replied:
Yes, exactly.

svyatonik commented on relays/lib-substrate-relay/src/finality/target.rs:
```diff
@@ -100,11 +100,11 @@ impl<P: SubstrateFinalitySyncPipeline, TargetClnt: Client<P::TargetChain>>
 		self.ensure_pallet_active().await?;
 
-		Ok(crate::messages::source::read_client_state::<P::TargetChain, P::SourceChain>(
+		Ok(best_synced_header_id::<P::SourceChain, P::TargetChain>(
```
It changes the logic a bit - previous call was returning the best available source header id at **best** target header and now it returns the best available source header id at **best finalized** target header. It is important, because when we are deciding - which source header to submit, we are looking at the **best** state of the target chain, not the **best finalized** state (otherwise our transaction will be obsolete).

serban300 replied:
I missed this. Fixed.

svyatonik commented on relays/utils/src/lib.rs:
```diff
@@ -134,12 +135,15 @@ pub enum TrackedTransactionStatus<BlockId> {
 	async fn wait(self) -> TrackedTransactionStatus<Self::HeaderId>;
 }
 
+pub type TrackedTransactionFuture<T> =
```
Please some docs for public type alias

serban300 replied:
Done.

svyatonik reviewed (commented):
Looks good - I've left some comments. Most are for the future PRs and ideas, so maybe we should just log it somewhere (some may be ignored I suppose :))

serban300 commented on relays/equivocation/src/reporter.rs:
```diff
@@ -0,0 +1,78 @@
+		at: P::Hash,
+		equivocation: P::EquivocationProof,
+	) -> Result<(), SC::Error> {
+		let pending_report = source_client.report_equivocation(at, equivocation).await?;
```
It looks like `report_equivocation()` [doesn't refund in case of duplicate equivocation reports](https://github.com/paritytech/substrate/blob/a732a9a7153b2be8424199b20e567e6358c25e17/frame/offences/src/lib.rs#L121). And it makes sense to be like this. Cause otherwise it looks like it could be a DOS vector.

And currently I don't think there is any way to check if any report was already submitted. Currently there doesn't seem to be any runtime API for getting this information. But it's a good point. Maybe we could add a runtime API method for this.

svyatonik replied:
Maybe we could read it from [here](https://github.com/paritytech/substrate/blob/a732a9a7153b2be8424199b20e567e6358c25e17/frame/offences/src/lib.rs#L73). In any case - it isn't a big deal, just some possible improvement :)

svyatonik replied:
> afaict valid reports are free even if duplicates: https://github.com/paritytech/substrate/blob/e806cefa1ebf0edfe493dafa184f7e50d8772a68/frame/grandpa/src/lib.rs#L212 (as long as they're valid)

If so, then anyone could just spam the network for free with `report_equivocation` transactions. Either it is true (and then the issue in in the `report_equivocation`), or it internally checks for duplicates and [returns error here](https://github.com/paritytech/substrate/blob/e806cefa1ebf0edfe493dafa184f7e50d8772a68/frame/grandpa/src/lib.rs#L210)

serban300 commented on relays/equivocation/src/reporter.rs:
```diff
@@ -0,0 +1,78 @@
+				Poll::Ready(tx_status) => {
+					match tx_status {
+						TrackedTransactionStatus::Lost => {
+							log::error!(target: "bridge", "Equivocation report tx was lost");
```
1. Totally agree. I was also planning this for a future PR.
2. Could you expand on this please ? I'm not sure what is the justifications db.

svyatonik replied:
2. I mean that in the future I think we should keep (store) all seen justifications in some database (something I've been talking about [here](https://github.com/paritytech/parity-bridges-common/pull/2348#discussion_r1294333348)). This would allow ED to outlive connection issues with target node (we are not simply losing justifications on reconnect/restart) + if we have failed to submit justification (don't know the reason - just imagine it has happened - e.g. we have failed to generate key ownership proof), we could then just get some alert (based on ED logs) and submit it manually using justifications db. Please note that justifications are not stored anywhere, so if we've missed it, it is just lost. And even if we see that something bad has happened, we have no means to submit equivocation proof - it requires a justification (actually a vote).

But that's a stuff for the future.

serban300 commented on relays/finality/src/finality_proofs.rs:
```diff
@@ -69,6 +69,12 @@ impl<P: FinalityPipeline, SC: SourceClientBase<P>> FinalityProofsStream<P, SC> {
 	}
 }
 
+impl<P: FinalityPipeline, SC: SourceClientBase<P>> Default for FinalityProofsStream<P, SC> {
+	fn default() -> Self {
+		Self::new()
+	}
+}
```
You're right. Done.

serban300 commented on relays/equivocation/src/reporter.rs:
```diff
@@ -0,0 +1,78 @@
+};
+
+pub struct EquivocationsReporter<P: EquivocationDetectionPipeline, SC: SourceClient<P>> {
+	pending_reports: Vec<TrackedTransactionFuture<SC::TransactionTracker>>,
```
I'll try it when implementing point 1. from [here](https://github.com/paritytech/parity-bridges-common/pull/2367/files#r1302616849). Might be easier to use this `Vec` if we also plan to store some extra data about the equivocations. Not sure.

serban300 commented:
@svyatonik @acatangiu thanks for the review. I addressed the comments and also created #2374 for tracking the things that need to be done in future PRs. PTAL when you have time.

svyatonik commented on relays/equivocation/src/equivocation_loop.rs:
```diff
@@ -0,0 +1,327 @@
+							continue
+						},
+						false => {
+							// Irrecoverable error. End the loop.
```
> we risk getting stuck in a retrial loop

That's true. Normally (in other relay loops) we'll eventually break of this loop when we'll read updated state and realized that something has changed. But until then we'll keep retrying. Which is fine (imo), unless we spam RPC server with failing RPC requests without some backoff.

In your examples:
- `we can't generate the key ownership prof, because we're too many blocks ahead`: then the loop should just log an error it and go further (imo);
- `let's say that the chains were updated and the relayer has old data structures, so it can't encode/decode them
etc`: that's definitely a maintenance flaw - the best we could do here is to have a backoff mechanism. Like your current loop implementation would just exit the process if you encounter any non-connection error. Is it good? E.g. if we can't generate ownership proof, then we could just keep working with next justifications.

But as I said - in the end it is up to you, how to handle issues :)

serban300 commented on relays/equivocation/src/equivocation_loop.rs:
```diff
@@ -0,0 +1,336 @@
+				let mut context =
+					match self.build_equivocation_reporting_context(current_block_number).await {
+						Some(context) => context,
+						None => return,
```
Done. Sorry, leftover from before, when I was treating all non-connection errors as unrecoverable

acatangiu replied:
With `continue` this becomes a busy-wait loop continuously retrying `try_read_from_target()` until `Ok(Some(_))`, with the `break` we'd at least get some ratelimiting from `tick`...

I guess both are fine if we expect `try_read_from_target()` to return `Ok(Some(_))` eventually...
